### PR TITLE
fix(core): Ensure execution deletion in worker lifecycle hook

### PR DESCRIPTION
**Pull request by ivov**
Reported by customer [here](https://n8nio.slack.com/archives/C05PUALKZHD/p1697446945481249?thread_ts=1697196557.638169&cid=C05PUALKZHD), apparently a very old long-standing bug for queue mode. Please review closely as I am not familiar with queue mode.

### CODE REVIEW

File: packages/cli/src/WorkflowExecuteAdditionalData.ts [hunk 1]
```diff
@@ -664,16 +664,39 @@ function hookFunctionsSaveWorker(): IWorkflowExecuteHooks {
...
+					let saveDataSuccessExecution = config.getEnv('executions.saveDataOnSuccess') as string;
+					if (this.workflowData.settings !== undefined) {
+						saveDataErrorExecution =
+							(this.workflowData.settings.saveDataErrorExecution as string) ||
+							saveDataErrorExecution;
+						saveDataSuccessExecution =
+							(this.workflowData.settings.saveDataSuccessExecution as string) ||
+							saveDataSuccessExecution;
+					}
+
```
Comment (ivov) [inline packages/cli/src/WorkflowExecuteAdditionalData.ts:678]: Duplication will be removed at https://github.com/n8n-io/n8n/pull/7370

File: packages/cli/src/WorkflowExecuteAdditionalData.ts [hunk 2]
```diff
@@ -664,16 +664,39 @@ function hookFunctionsSaveWorker(): IWorkflowExecuteHooks {
...
+								this.workflowData,
+								fullRunData,
+								this.mode,
+								this.executionId,
+								this.retryOf,
+							);
+							await Container.get(ExecutionRepository).hardDelete({
+								workflowId: this.workflowData.id as string,
+								executionId: this.executionId,
+							});
```
Comment (krynble) [inline packages/cli/src/WorkflowExecuteAdditionalData.ts:696]: Unfortunately this deletion cannot happen here.

Due to the nature of queue mode, we save the execution data to DB and notify via Redis that a given execution has finished.

This in turn allows the caller (be it `main` or `webhook`) to load the execution from DB (if needed) and react (such as sending an HTTP response or removing an item from a queue).

This code would need to be part of the array returned by `getWorkflowHooksWorkerMain` instead.

**Review by krynble — CHANGES_REQUESTED** [review]

File: packages/cli/src/WorkflowExecuteAdditionalData.ts [hunk 3]
```diff
@@ -1130,6 +1130,36 @@ export function getWorkflowHooksWorkerMain(
...
+			// Check config to know if execution should be saved or not
+			let saveDataErrorExecution = config.getEnv('executions.saveDataOnError') as string;
+			let saveDataSuccessExecution = config.getEnv('executions.saveDataOnSuccess') as string;
+			if (this.workflowData.settings !== undefined) {
+				saveDataErrorExecution =
+					(this.workflowData.settings.saveDataErrorExecution as string) || saveDataErrorExecution;
+				saveDataSuccessExecution =
+					(this.workflowData.settings.saveDataSuccessExecution as string) ||
+					saveDataSuccessExecution;
+			}
```
Comment (ivov) [inline packages/cli/src/WorkflowExecuteAdditionalData.ts:1148]: Duplication will be removed at #7370

**Review by krynble — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.14.2`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.14.2)

**Pull request merged**